psi4/src/psi4/dlpno/mp2.h

```diff
@@ -0,0 +1,180 @@
+ */
+
+#ifndef PSI4_SRC_DLPNO_MP2_H_
+#define PSI4_SRC_DLPNO_MP2_H_
+
+#include "sparse.h"
+
+#include "psi4/libmints/wavefunction.h"
+
+namespace psi {
+namespace dlpno {
+
+// Equations refer to Pinski et al. (JCP 143, 034108, 2015; DOI: 10.1063/1.4926879)
+
+class DLPNOMP2 : public Wavefunction {
+   protected:
+
+    /// threshold for PAO domain size
+    double T_CUT_DO;
+
+    /// threshold for PNO truncation
+    double T_CUT_PNO;
+
+    /// auxiliary basis
+    std::shared_ptr<BasisSet> ribasis_;
+    SharedMatrix full_metric;
+
+    /// localized molecular orbitals (LMOs)
+    SharedMatrix C_lmo;
+    SharedMatrix F_lmo;
+
+    /// projected atomic orbitals (PAOs)
+    SharedMatrix C_pao;
+    SharedMatrix F_pao;
+    SharedMatrix S_pao;
+
+    /// differential overlap integrals (EQ 4)
+    SharedMatrix DOI_ij; // LMO/LMO
+    SharedMatrix DOI_iu; // LMO/PAO
+
+    // approximate LMO/LMO pair energies from dipole integrals (EQ 17)
+    SharedMatrix e_actual; ///< actual approximate pair energy
+    SharedMatrix e_linear; ///< upper bound to approximate pair energies
+
+    /// LMO/PAO three-index integrals
+    std::vector<SharedMatrix> qia;
+
+    /// pair natural orbitals (PNOs)
+    std::vector<SharedMatrix> K_iajb;  ///< exchange operators (i.e. (ia|jb) integrals)
+    std::vector<SharedMatrix> T_iajb;  ///< amplitudes
+    std::vector<SharedMatrix> Tt_iajb; ///< antisymmetrized amplitudes
+    std::vector<SharedMatrix> R_iajb;  ///< LMP2 residuals
+    std::vector<SharedMatrix> X_pno;   ///< global PAO -> canonical PNO transforms
+    std::vector<SharedVector> e_pno;   ///< PNO orbital energies
+    std::vector<int> n_pno;       ///< number of pnos
+    std::vector<double> de_pno;   ///< PNO truncation energy error
+    std::vector<std::vector<SharedMatrix>> S_pno_ij_kj; ///< pno overlaps
+    std::vector<std::vector<SharedMatrix>> S_pno_ij_ik; ///< pnooverlaps
+
+    // final energies
+    double de_dipole_; ///< energy correction for distant (LMO, LMO) pairs
+    double de_pno_total_; ///< energy correction for PNO truncation
+    double e_lmp2_; ///< raw (uncorrected) local MP2 correlation energy
+
+    // => Sparse Maps <= //
+
+    // orbital / aux bases
+    SparseMap atom_to_bf_; ///< which orbital BFs are on a given atom?
+    SparseMap atom_to_ribf_; ///< which aux BFs are on a given atom?
+    SparseMap atom_to_shell_; ///< which orbital basis shells are on a given atom?
+    SparseMap atom_to_rishell_; ///< which aux basis shells are on a given atom?
+
+    // LMO domains
+    SparseMap lmo_to_ribfs_; ///< which aux BFs are needed for density-fitting a LMO?
+    SparseMap lmo_to_riatoms_; ///< aux BFs on which atoms are needed for density-fitting a LMO?
+    SparseMap lmo_to_paos_; ///< which PAOs span the virtual space of a LMO?
+    SparseMap lmo_to_paoatoms_; ///< PAOs on which atoms span the virtual space of a LMO?
+    std::vector<std::vector<int>> i_j_to_ij; ///< LMO indices (i, j) to LMO pair index (ij)
```

Comment:
Please include a comment about how you are indexing pairs of _significant_ pairs, and that an insignificant (i, j) hashes to -1.